### DISCUSSION TITLE
[Bug]: Breaking SOLID principles and Clean Architecture.

### DISCUSSION THREAD

**Discussion by motornyimaksym** [android/nowinandroid#1273] [General]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is there a StackOverflow question about this issue?

- [X] I have searched StackOverflow

### What happened?

Somebody sent me your project as a best example of Clean Architecture and the only one right reference from Google for those who study programming (meaning that this is the only one unbreakable existing source of truth of architecture).
I open the dependencies and see that domain layer depends on data. Which has no sense if your definition of the domain is the same as canonical definition of pure business logic layer. 

It would be nice to add a disclaimer that the domain layer from this project has nothing in common with Clean Architecture's domain except the title. Or you may use other title for your layer because the domain logic can not depends on the data source implementation. E.g. "logic" layer, but not domain. It's very confusing because you substitute widely known definition of the domain layer with your own one and it gives wrong direction for those who can't argue against the huge Google authority. 

On other hand you may slightly refactor the project by changing the way of dependencies so the data layer will depends on domain layer. In this case you'd need to define repository interfaces inside of the domain module and create their implementation inside of the data layer. It will cost almost nothing in terms of implementation efforts but will bring much more clarity for those who try to google question like "data layer depends on domain". 

[People already started asking questions](https://stackoverflow.com/questions/77337477/android-multi-module-architecture-domain-depends-on-data-or-vice-versa).

In my opinion it's a major architectural issue because the data layer implementation should depends on abstract domain logic (including abstract repository interface) but not vice versa. Without doubts it's up to you keep everything as is but in this case I'd kindly ask you to add disclaimer directly to this project readme file with some clarifications and explanations.

Looking forward for your thoughts. Thanks in advance.

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by SimonMarquis** [comment]
_Disclaimer: I'm not affiliated with Google in any way._

Thanks for raising your concerns (that I personally share as well), but I don't think Google ever said they were following "the clean architecture". Not on their [official documentation](https://developer.android.com/topic/architecture/intro), not on this project either.  
I understand that re-using the same terms can be confusing, but there are not a lot of choice, and naming things is hard. It's like saying MVVM can't use the terms "View" or "Models" because MVP already used them before. 😛
They could probably add a disclaimer somewhere [in here](https://developer.android.com/topic/architecture/domain-layer) though.
But keep in mind that this is just one way of doing architecture, and Google does not force anyone to use it as-is.
That being said, this project is probably not the best place to make your voice heard, as this repository is only reflecting what the official Google recommendations are at this time. Providing feedback or participating in official Q&A might be more effective.
If the maintainers of this project could clarify the situation, it would be great.

**Comment by dturner** [comment]
Thanks for raising this. Let me try to provide some clarity. I'm an engineer on the Android DevRel team at Google and I'm the Tech Lead on this project.

**Why doesn't Now in Android use "Clean Architecture?"**

As @SimonMarquis said, Now in Android is intended to be a real-world implementation of the [official Architecture guidance](https://developer.android.com/topic/architecture/intro) (as well as many other best practices for testing, UI design etc). As you have pointed out, the official guidance is different from "Clean Architecture" in two fundamental ways. In the official guidance: 

- the domain layer is optional
- the domain layer depends on the data layer, not the other way round

**Why doesn't Google just recommend "Clean Architecture"?**

Many apps aren't complex enough to justify having a domain layer. This is why in the official guidance the domain layer is optional. You can build your app with the UI layer dependent on the data layer, then introduce the domain layer later if the need arises. Ultimately we feel that the official guidance offers the best approach for most Android apps. 

It's worth saying that the official guidance is just that: guidance. It's not intended to be a one-size fits all. Developers should question whether the ui->(optional domain)->data approach is right for their app's requirements.

**What are the benefits of the official Android guidance over Clean Architecture?**

- Less rigid. You can start off with just UI and data layers, introducing the domain layer only when it's required for your app to scale.
- Easier to change where data is stored (e.g. by moving from preferences to a Room database or local to cloud) since the data layer isn't dependent on any other layers.

**What are the benefits of Clean Architecture over the official Android guidance?** 

- Easier to change business logic. Clear encapsulation of this logic from the outset. No business logic in UI or data layers. 
- In Android apps strictly following Clean Architecture, the business logic is usually contained in plain JVM library modules (which have no Android dependency). This can be useful in cross-platform apps.

Hopefully that answers some of your questions. 

One question in return (to @Yazon2006): you mentioned that Now in Android is breaking SOLID principles - where? I'm not saying it's perfect but I don't see any glaring violations of SOLID. 

Happy to discuss further.

**Reply by motornyimaksym** [reply]
> Now in Android is breaking SOLID principles - where

I already explained that the problem in definition substitution and agree with @SimonMarquis that this is the problem not of this exact project but the [official Architecture guidance](https://developer.android.com/topic/architecture/intro). Domain layer in other architecture descriptions (e.g. Clean Architecture + Domain Driven Development concept) means the layer with the business rules and domain objects. Business rules can not depends on what exact library we use for storage or which exact server we call for getting the information. So in a nutshell domain is an abstraction. Data layer in our example and any other architecture descriptions contains pretty exact implementations that use exact libraries etc. According to the fifth SOLID principle abstractions should not depend on details; details (concrete implementations) should depend on abstractions.

In [official Architecture guidance](https://developer.android.com/topic/architecture/intro) the defitnition of [domain layer](https://developer.android.com/topic/architecture/domain-layer) looks more not like additional architectural layer but the description of the [facade](https://refactoring.guru/design-patterns/facade) pattern that potentially can be used for creating simplified interface of complex set of classes.

I clearly remember the previous variant of recommended app architecture:
![image](https://github.com/android/nowinandroid/assets/3704058/3d2de21a-3f81-4851-ada1-86e5ac144caf)
And in the case above we see that we can add optional facade class that will provide access to few repositories and maybe not only repositories other mapper/utility classes. It will not require any additional architecture layer in meaning of gradle module.

> Many apps aren't complex enough to justify having a domain layer.
Let's be honest. Many apps aren't complex enough to justify having [official Architecture](https://developer.android.com/topic/architecture/intro) in a way that it's represented in Now in Android app. Everything can be implemented in single gradle module without having discussions why domain depends on data. 

And what is most important is that the UseCase/Interactor is optional in Clean Architecture as well as any other additional layer of abstraction. I know that a lot of devs are tired to create proxy pattern based UseCases just to call repository from ViewModel but it's not Clean Architecture issue. It's a problem of wrong understanding of it. We can easily use repository interface, which belongs to the domain layer, to access the data directly in ViewModel. What can't be optional is the busines entity model scheme (e.g. [UserData](https://github.com/android/nowinandroid/blob/main/core/model/src/main/kotlin/com/google/samples/apps/nowinandroid/core/model/data/UserData.kt)) which in our particular case belongs to **core** module/layer that even has no description in [official Architecture guidance](https://developer.android.com/topic/architecture/intro). In CA that model would be a part of the domain (or even lower optional layer with only models, e.g. core/entities) and data would depends on domain providing the repository that follow interfaces contains enitty models.

And now imagine trainee/junior or even middle dev who are just trying to make their own project. They google "android app architecure", find [official Architecture guidance](https://developer.android.com/topic/architecture/intro), find [Now in Android](https://github.com/android/nowinandroid) example and **blindly** follow the recommendations because they are not experienced enough to take what is usefull and abandon what is not. In this case there could be two possible outcomes: 1) if the project is small enough then the recommended architecture would be much more complicated then they required 2) if the project is big/complex enough to require scalable/reliable and widely recognized architecture then it's better to follow CA and at least have independent domain layer with entities/interfaces/relations/etc.

I agree that my concerns more related to the architecture itself than the Now in Android project. However this is not the worst place to discuss it. Thanks a lot your answers and thoughts.

**Reply by arkivanov** [reply]
I think it's arguable whether a domain-layer _class_ A that depends on a data-layer _interface_ B from a module that also contains an implementation of B is a violation of the "D" principle or not. The concrete class A still depends on an abstraction B, even though there BImpl next to B. Looks fine from my point of view, and also flexible enough as the implementation can be easily moved to another module when needed (e.g. we need another implementation of B).

**Reply by arkivanov** [reply]
On the other hand, why would a data layer module depend on a domain layer module, given that implementations from the data module don't need all those use cases from the domain module?

**Reply by motornyimaksym** [reply]
@arkivanov [High-level modules should not import anything from low-level modules](https://en.wikipedia.org/wiki/Dependency_inversion_principle). [Domain](https://en.wikipedia.org/wiki/Domain_(software_engineering)) in our case is high level module and data is low level module. When domain module depends on data it's possible to break this rule intended or not. When data depends on domain such situation (when domain imports something from data) not possible at all. The whole purpose of having all those gradle modules is separation of concerns otherwise we may put everyting as is into the one gradle module and keep following only part B of dependency inversion principle like you described.

**Reply by motornyimaksym** [reply]
> why would a data layer module depend on a domain layer module

I explained it in details here:
https://github.com/android/nowinandroid/discussions/1273#discussioncomment-8766215
Short answer is because the domain layer in case of this repo mainly represented by `core.model`

**Reply by arkivanov** [reply]
I disagree. From the same "Dependency Inversion" wiki link:

> thus rendering high-level modules independent of the low-level module implementation details

This is exactly how NowInAndroid is built (as per my understanding). Classes from the domain module do not import implementation details from the data module.

Also the generic term "module" doesn't imply "Gradle module" - a module can be a class. The fact that the data module contains both kind of entities - interfaces and implementation details - doesn't break any SOLID rules. _As long_ as entities from the domain module do not import implementation details from the data module. The location on disk, folder structure, Gradle modularity don't matter much. It's possible to just split the `data` Gradle module by two - `data-api` and `data-impl` - and let the `domain` module to only depend on `data-api`. But this doesn't change anything much.

**Reply by Jaehwa-Noh** [reply]
Oh, It seemed to matter of definition of module.

**Reply by motornyimaksym** [reply]
> "module" doesn't imply "Gradle module"

As well as term "import" doesn't imply "java class import. Think of it in other way. Domain layer transitively add dependency on data module which contains database, datastore, network. It doesn't makes sense because the logic in usecase do not depends on use we ktor or retrofit. 

> But this doesn't change anything much.

This is dramatically changes everything. In case if we split data into `data-api` and `data-impl` so that the 
- `data-api` will contain only interfaces and depends on `core.model`
- `core.domain` will depends on `data-api` and `core.model`

then the `core.model` + `core.domain` + `data-api` will effectively form the domain layer in it's classic definition and the `data-impl` will depends on part of it. That's the same as "make data depends on domain" but even more precise in terms of separation of concerns. 

The simplest example of domain layer is when we have models + interfaces + usaceses in one module. But it's dangerous because in growing project the complexity of this module will lineary grow and become most complex non scalable part. With `-api`/`-impl`/etc modules concept we always can extract piece of domain knowledge into separate `-api` feature module, reuse it in other feature without exact implementation. We can go even further and extract `-model` into separate module. But there should be always the same rule: impl will depends on api, not wice versa. 

> As long as entities from the domain module do not import implementation details from the data module

How many trainee/junior and even middle devs can break this rule using current dependency direction setup? I remind you that this is an example from Google. It's a Holy Scripts of android development community. Can they resist temptation to import retrofit dto model into usecase to avoid creating dto->domain mapper? It's absolutely easy in this project. We can even avoid creating model in `core.model` and use only publicly available dto model in all data-domain-presentation layers.

**Reply by briandr97** [reply]
> **What are the benefits of the official Android guidance over Clean Architecture?**
Easier to change where data is stored (e.g. by moving from preferences to a Room database or local to cloud) since the data layer isn't dependent on any other layers.

@dturner I disagree with this: why do you think it's harder to change where data is stored in a Clean Architecture? I think it's still easy to change data source(where data is stored) in the data layer when we use Clean Architecture
Whatever changes happen in the data layer, the data layer only needs to implement the interfaces of the domain layer. Therefore, changes in the data layer don't propagate to other layers.

However In Android guidance, data layer is purely independent and other layers such as domain, ui should depend on data layer. So data layer should be stable. It is difficult to change a software component that many other components depend on it.

You might say that since an interface exists between the data layer and the other layers, you can change data source without changing the interface. 
Even taking that into account, it cannot be said that CA is more harder to change where data is stored than Android guidance.

**Reply by dturner** [reply]
Yes, you're right to challenge that point. If you follow good design practices it shouldn't be any easier to change the data layer in the official guidance vs Clean Architecture. The difference occurs if you have a naively designed data layer which does not use interfaces and instead relies on concrete implementations. In this scenario it _would_ be easier to change the data layer because your implementations are not dependent on the domain layer.

**Reply by briandr97** [reply]
Thanks for your response.
I agree that if an interface exists between layers, there is not much difference between the two architectures in terms of changing where data is stored.
However, your assumption of no interfaces is difficult to understand. Because in CA, dependency inversion through interfaces is essential.
I think I would agree if you are saying that CA has a higher barrier to entry than Google guidance because it requires an understanding of inter-layer dependency inversion.

**Comment by Jaehwa-Noh** [comment]
Wait... Are they clean architecture and Google's architecture guideline different?
I have another question.

I saw this two images.

<img src="https://github.com/android/nowinandroid/assets/48680511/22119967-e152-4b55-a96c-8dacaa134144" width="400"/>   

[figure 1. Clean architecture]

<img src="https://github.com/android/nowinandroid/assets/48680511/d9290164-3582-4ba1-bd63-5afbe122b159" width="300"/>   

[figure 2. Google's architecture guideline]




Both look similarly same to me.
Let's rotate figure 2.


<img src="https://github.com/android/nowinandroid/assets/48680511/6bfa36e9-422c-4629-ad76-a8624bf4da6c" width="150"/>   

[figure 3. Google's architecture guideline(Sorry for rotating)]

figure 3 looks like figure 1.


I thought I can connect each others architecture's.
Clean architecture -> Google's architecture guideline
• UI, Presenters -> UI Layer
• Use Cases -> Domain Layer
• Entities -> Data Layer

What do you think about my idea?
If both are different, what did I miss the point of view?

Thank you.

**Reply by dturner** [reply]
Sorry, I accidentally started a new thread below instead of replying directly.

**Reply by Jaehwa-Noh** [reply]
It's ok. I can feel that reply was for me.

**Comment by dturner** [comment]
You're right, they are similar, but not identical. The differences are highlighted in red below: 

<img width="1313" alt="image" src="https://github.com/android/nowinandroid/assets/873212/8978b770-ca9c-46fc-83e0-08da73574cb5">

As an exercise, it could be interesting to migrate the existing architecture to clean architecture. It should be a relatively simple refactor to move the `:core:data` repository interfaces into the `:core:domain` module, reversing the dependency.  

For strict adherence to clean architecture, you'd also need to write wrapper functions in the domain layer to remove the ui->data layer dependency

**Reply by Jaehwa-Noh** [reply]
I see! Thank you for your kindness.

**Reply by motornyimaksym** [reply]
> For strict adherence to clean architecture, you'd also need to write wrapper functions in the domain layer to remove the ui->data layer dependency

For pure android projects it has no sense at all. Ui and data are so coupled that trying to isolate them will just make the dependency transitive and bring more boilerplate code without any benefits in practice.

**Reply by dturner** [reply]
That's a good point. I don't have much practice with Clean Architecture and it makes perfect sense that once the repository interfaces are in the domain layer you can call them directly.

**Reply by arkivanov** [reply]
From my point of view having repository interfaces in the domain layer is not good. E.g. if I have use case classes and repository interfaces used by those use case classes in the domain layer, then why would my data layer that depends on the domain be aware of the use case classes? It doesn't make sense.

The way the "Official Android guidance" recommends looks much better to me. If really needed, we can split the data layer by two - data-api and data-impl, and let the domain layer depend on data-api. So the data-api thing contains only API for data access, which is exactly what is required in the domain layer.

```mermaid
graph TD;
    UI-->Domain;
    Domain-->Data-Api;
    Data-Impl-->Data-Api;
```

**Reply by dturner** [reply]
Yeah, my point was that _if you are adopting Clean Architecture_ where it is mandated that the repository interfaces reside in the domain layer you don't need to provide UseCases for all repository methods.

I agree that having data repository interfaces in the domain layer feels kind of strange, however, I guess the point is that in Clean Arch you are prioritising business logic over everything else. The repository interfaces represent the contract which all data providers must conform to, and therefore should reside in the domain layer. 

To me, this is splitting hairs though, as long as you are actually splitting interfaces and implementations, whether that's: 

- within the same module (which is what we do in NiA e.g. `:core:data`)
- separate `api` and `impl` modules (good suggestion)
- or in a domain module (Clean arch).

it's actually not that important. What is important is that people understand the project architecture and use it correctly. 

I will say that in my (extremely biased) opinion (and this is backed up by your comment @arkivanov) the official architecture is easier to grasp from a conceptual level, simply because the dependency arrows all go the same way.

**Reply by motornyimaksym** [reply]
> From my point of view having repository interfaces in the domain layer is not good.

So then lets follow [Separation Of Concerns](https://en.wikipedia.org/wiki/Separation_of_concerns) principle as stated in [docs](https://developer.android.com/topic/architecture)

Define domain with definitions and relations:

Models aka simple pojo classes for domain entities. E.g. Car, Human, etc. Then relations. Human that drive the car is Driver.  Than you define the abstractions for data sources for Cars, Humans. E.g. buyCar(money: Money): Car; hireDriver(money: Money). Then some logic getCars+getHumans=makeTheRace. That's all about the domain. If you wan't you may split it in as many sublayers as you want. E.g. in this repo `core.model` could be the bottom part, `data-api` somewhere between `core.model` and `business-logic`. And the `business-logic` will include two previous.
 
Now we can proceed with implementation:

We depends on everything that I described above. We may have different features that use any described particle `model`/`data-api`/`domain-logic`. And indeed everything above the model level is optional in terms of some particular feature needs. Let's say I develop advertisement feature. I don't care about data sources or logic, I only need models. My feature will use models to create slogan: "Best Cars for Humans". That's it. No necessity to depends on `data-api`/`business-logic`. Or I may wan't to create some specific feature with some specific logic that depends only on data source, e.g. getHumans+getCars=makeTheExhibition. Then it obviously will depends on `model`/`data-api` without `business-logic`. Or just some simple UI mostly module that will consume the `business-logic` including `model`/`data-api`.

Then it will increase the complexity of the project sctructure but bring more separation of concerns.
In case if we want to go back to simplify the project we may combine model+data-api+business-logic into one big domain layer and  everything would be fine except the scalability problem. That's always a trade off.

And all of that starting from basic models and up to the place where we start talking about UI that's the domain. The domain definition is not limited only to UseCases. That is what I'm trying to explain from the very beginning. The definition of the domain in [documentation](https://developer.android.com/topic/architecture) is wrong. The domain can't be optional while usecase can be. 

Rename the "domain" layer in docs to "business-logic" and you'll get traditional layered architecture:
![image](https://github.com/android/nowinandroid/assets/3704058/56c63eba-c924-4b7c-bf76-85309f4908db)
(Hello 2000)
Where the business logic depends on the exact data source. It's when you need to rewrite UseCase as soon as you would need to migrate your app from home ftp server to Amazon AWS. It's when migration from Retrofit to Ktor require domain layer changes. It's when the definition of Car depends on exact shop you buy it from. 

> the official architecture is easier to grasp from a conceptual level, simply because the dependency arrows all go the same way

I have nothing against this approach. It still widely used. But if google recommend something then it would be really kind from their side to let users know about disadvantages of such approach at least.

We obviously see that we may change the direction of the injection and get independent "domain" layer. Where the situation when the logic depends on exact implementation would be not possible.

For experienced developer this repository is just another example of architecture. We may take what we wan't from here and forget about the rest.

But for those who study, for those who has no strong opinion this repo is the single source of truth. They'll take this project as a barebone for their pet projects. And will build interfaces and logic depends on the exact data sources. And then they will came to all of us on the interview with examples of how they thought Clean Architecture just because it has domain layer =). It's not CA example and will not be after proposed changes. Moreover as I said I don't like CA because it's complicated. But this project is not much easier than CA for understanding. So we will not takle all the problems at once. But the dependency direction rule change is a low hanging fruit that without doubts worth to pick up.

**Reply by dturner** [reply]
> The definition of the domain in [documentation](https://developer.android.com/topic/architecture) is wrong.

It's only wrong if you take it out of context and assume that it applies to Clean Architecture. It isn't wrong in the context of the official Android architecture guidance. 

> Where the business logic depends on the exact data source. It's when you need to rewrite UseCase as soon as you would need to migrate your app from home ftp server to Amazon AWS. It's when migration from Retrofit to Ktor require domain layer changes. It's when the definition of Car depends on exact shop you buy it from.

No-one is suggesting this and this is not how Now in Android is implemented. 

The data layer interfaces represent the public contract with the domain layer, not the underlying implementations. For example, in Now in Android you could change from Retrofit to Ktor simply by adding a `KtorDataSource` which implements [`NiaNetworkDataSource`](https://github.com/android/nowinandroid/blob/main/core/network/src/main/kotlin/com/google/samples/apps/nowinandroid/core/network/NiaNetworkDataSource.kt). No UseCases or repository implementations would change.

To be clear, we are still using (and recommending) dependency inversion (as explained in the [modularization section of the official architecture guidance](https://developer.android.com/topic/modularization/patterns#dependency_inversion)). It's just applied at a module/object level, rather than at an architectural layer level.

**Reply by motornyimaksym** [reply]
@dturner Ok, I got your point. I already said everything I had. If you insist that the definition of domain in your docs is ok then so be it. If you think that domain dependency on data is ok I'm fine with it. Just please, add some **disclaimer** in docs and into repo readme. Something like: "**Our definition of "domain" is different from the definition of "domain" that is used in Clean Architecture by Robert C. Martin or/and Onion Architecture by Jeffrey Palermo**". That's all what I ask. Hope it makes sense. 
P.S. Such disclaimer can explain most of architectural peculiarities.

**Reply by dturner** [reply]
Added a note to the architecture docs for the project (will merge once approved) https://github.com/android/nowinandroid/pull/1304

**Comment by dturner** [comment]
I decided to see how easy it is in reality to refactor to clean architecture. Summary: around 1 hour's work. https://github.com/android/nowinandroid/pull/1277

**Reply by motornyimaksym** [reply]
I precisely reviewed your PR. Looks as good as simple. That's the main step from the traditional to the onion architecture which is good described in the article [The Onion Architecture](https://jeffreypalermo.com/2008/07/the-onion-architecture-part-1/) from [Jeffrey Palermo](https://jeffreypalermo.com/author/jeffreypalermo/) (year 2008, what a lovely year). So now we are more close to the classic definition of the domain layer that was described in the the Onion Architecture (the father of CA). 

A small step for [official Architecture guidance](https://developer.android.com/topic/architecture/intro) a giant leap for developers community.

I'd like also to mention that the domain layer can't be optional because first of all it is represented by `core.model` module that contains pure domain models (once again I don't see core layer in [official architecture](https://developer.android.com/topic/architecture)). This is actually the very bottom of the layered onion architecture. This is the true domain. The area of project wide (not only android but backend+other platforms) definitons. The [Topic](https://github.com/android/nowinandroid/blob/main/core/model/src/main/kotlin/com/google/samples/apps/nowinandroid/core/model/data/Topic.kt) entity description would remain the same regardless what database we use on what platfrom etc. [FollowableTopic](https://github.com/android/nowinandroid/blob/main/core/model/src/main/kotlin/com/google/samples/apps/nowinandroid/core/model/data/FollowableTopic.kt) is an entity that describes some busines logic. It show the relation between the topic and some boolean state that represents if this topic is followed or not. [GetSearchContentsUseCase](https://github.com/android/nowinandroid/blob/main/core/domain/src/main/kotlin/com/google/samples/apps/nowinandroid/core/domain/GetSearchContentsUseCase.kt) goes further and establish a bunch of relations between FollowableTopic, UserData, UserNewsResource inside of UserSearchResult. At this point we have such a lot of beautiful abstract sets of definitions, relations and rules. And there are no any word about exact database or some particular remote service. Android+backend+ios+anyOther would speak with each other using the same language (same definitions) described in our domain layer without knowing low level details. Isn't it a much better aproach?
This is exactly what [fifth SOLID principle](https://en.wikipedia.org/wiki/Dependency_inversion_principle) tells us. High-level modules (domain) should not import anything from low-level modules (module with retrofit or sqlite). Both should depend on abstractions (e.g., interfaces). Abstractions should not depend on details.

UseCases still can be optional because not all of them equaly useful. Especially talking into account examples like [GetSearchContentsCountUseCase](https://github.com/android/nowinandroid/blob/main/core/domain/src/main/kotlin/com/google/samples/apps/nowinandroid/core/domain/GetSearchContentsCountUseCase.kt) which is useless proxy class. Besides that UseCases could provide their functionality throught the interfaces to follow solid to the end. But for the sake of simplicity I suggest to do it only in case if they are used somewhere in other interface. 

In case if in viewModel we'll use repository interface directly without usecase to access the data we still use domain dependency because the repository interface now in the domain layer. That mean we even have some sort of UI separation from data. Which is not required for pure android projects but why not.
I found proposed PR very positive in terms of common understading of good architecture and simple enough to be adopted.
On other hand do you have any reason not to merge it except the necessity to rewrite the [official Architecture guidance](https://developer.android.com/topic/architecture/intro)?

**Reply by dturner** [reply]
Thanks for the review, and the feedback. I understand the strong feelings around this topic and will ensure that these points are raised with the team with a view to updating the guidance in future (and subsequently our samples). 

> `GetSearchContentsCountUseCase` which is useless proxy class

Agreed. This should be refactored to call the repository interface directly. Issue filed https://github.com/android/nowinandroid/issues/1283

> On other hand do you have any reason not to merge it except the necessity to rewrite the [official Architecture guidance](https://developer.android.com/topic/architecture/intro)?

We cannot diverge from the official guidance I'm afraid, that's one of the fundamental principles of this project.

**Reply by motornyimaksym** [reply]
> one of the fundamental principles

But it's always good to think about revisiting principles =) We are here to change the world to better

**Comment by anlexN** [comment]
This is a good talking.

clean architecture's diagram is circle, however android architecture's diagram is layered.
clean direction is two way, however android direction is one way.

some more are different.

So I want to say that if angular and android code architecture can be the same, it will be amazing.
